pallets/committee/src/lib.rs

```diff
@@ -240,6 +238,21 @@ pub mod pallet {
 			}
 			0
 		}
+
+		fn on_finalize(n: BlockNumberFor<T>) {
+			let voting_period = T::VotingPeriod::get();
+			Proposals::translate_values(|mut proposal: Proposal<T>| {
+				Self::get_votes_for(&proposal.hash()).and_then(
+					|votes: VoteAggregate<T::AccountId, T::BlockNumber>| -> Option<Proposal<T>> {
+						if n.saturating_sub(votes.end) > voting_period && proposal.status == ProposalStatus::Active {
+							proposal.status = ProposalStatus::Closed;
+						}
+
+						Some(proposal)
+					},
+				)
+			})
+		}
 	}
```

Comment:
This is a hook to check timed proposals, right?
it probably would be more efficient to only run this once per voting period?
also what does `translate_values` do exactly?
Wouldn't it be more efficien to keep the `ExecutedProposals` map and loop over the `Proposals` map here? this way we could get rid of the ProposalState

Reply:
> it probably would be more efficient to only run this once per voting period?

There are different voting period for different proposals, for example, as we set `VotingPeriod` to `10`, Alice submits a proposal at block 1, the voting period is `1~11`, Bob submits a proposal at block 3, the voting period will be `3~13`, we could not run this once per voting period because we don't we when people submit proposals

> also what does translate_values do exactly?

https://substrate.dev/rustdocs/latest/frame_support/storage/types/struct.StorageMap.html#method.translate_values

> Wouldn't it be more efficient to keep the ExecutedProposals map and loop over the Proposals map here? this way we could get rid of the ProposalState

If keep `ExecutedProposals`, there will need another storage map `ClosedProposals` or `Timeout?Proposals` because the expired proposals are not executed, otherwise we can modify `ExecutedProposals` to `ClosedProposals`.

For why we need to make ExecutedProposals and `ClosedProposals` different, plz take a look at https://polkadot.polkassembly.io/

Reply:
Or execute this in the `close` function, if expired close but not execute

Reply:
> > it probably would be more efficient to only run this once per voting period?
> 
> From the exist implementation, there are different voting periods for different proposals, for example, as we set `VotingPeriod` to `10`, Alice submits a proposal at block 1, the voting period is `1~11`, Bob submits a proposal at block 3, the voting period will be `3~13`, we could not run this once per voting period because we don't know when people submit proposals

I see, makes sense.

> > also what does translate_values do exactly?
> 
> https://substrate.dev/rustdocs/latest/frame_support/storage/types/struct.StorageMap.html#method.translate_values
> 
> > Wouldn't it be more efficient to keep the ExecutedProposals map and loop over the Proposals map here? this way we could get rid of the ProposalState
> 
> If keep `ExecutedProposals`, there will need another storage map `ClosedProposals` or `Timeout?Proposals` because the expired proposals are not executed, otherwise we can modify `ExecutedProposals` into `ClosedProposals` directly.
> 
> For why we need to make `ExecutedProposals` and `ClosedProposals` different, plz take a look at https://polkadot.polkassembly.io/


Do we really need to close them automatically? Since we know that they're timed out we could instead fail on `close`, this would simplify things quite a bit. The the state of the proposal makes perfect sense.

Reply:
Closing them automatically cost a lot and really dirty indeed, but imaging a UI of PINT, people query status of proposals, they'll get `Active` for proposals which are timeout when the committee haven't close them by hand.

> Since we know that they're timed out we could instead fail on close, this would simplify things quite a bit

I'm considering what about we close timeout proposals without throwing errors in the close function?

```
if proposal is timeout
  mark_timetout();
else
  execute() && mark_executed()
```